client/consensus/common/src/lib.rs

```diff
@@ -147,6 +147,11 @@ where
 		let hash = params.post_hash();
 		let number = *params.header.number();
 
+		if params.with_state() {
+			// Force imported state finality
+			params.finalized = true;
```

Comment:
It would be nice to have some tests for it, but I also don't have any good idea where to put it.

Why did you decided against the approach of doing this in `WarpSyncEngine` in Substrate?